test/pylib/s3_server_mock.py

```diff
@@ -0,0 +1,206 @@
+
+    def __init__(self, policies, logger, *args, **kwargs):
+        self.policies = policies
+        self.logger = logger
+        super().__init__(*args, **kwargs)
+        self.close_connection = False
+
+    def log_error(self, format, *args):
+        if self.logger:
+            self.logger.info("%s - - [%s] %s\n" %
+                             (self.client_address[0],
+                              self.log_date_time_string(),
+                              format % args))
+        else:
+            sys.stderr.write("%s - - [%s] %s\n" %
+                             (self.address_string(),
+                              self.log_date_time_string(),
+                              format % args))
+
+    def log_message(self, format, *args):
+        # Just don't be too verbose
+        if not self.logger:
+            sys.stderr.write("%s - - [%s] %s\n" %
+                             (self.address_string(),
+                              self.log_date_time_string(),
+                              format % args))
+
+    def parsed_qs(self):
+        parsed_url = urlparse(self.path)
+        query_components = parse_qs(parsed_url.query)
+        for key in parsed_url.query.split('&'):
+            if '=' not in key:
+                query_components[key] = ['']
+        return query_components
+
+    def do_POST(self):
+        content_length = self.headers['Content-Length']
+        if content_length:
+            put_data = self.rfile.read(int(content_length))
+        self.send_response(200)
+        self.send_header('Content-Type', 'text/plain; charset=utf-8')
+        self.send_header('Connection', 'keep-alive')
+        response = self.build_POST_reponse(self.parsed_qs(), urlparse(self.path).path).encode('utf-8')
+        self.send_header('Content-Length', str(len(response)))
+        self.end_headers()
+        self.wfile.write(response)
+
+    def do_PUT(self):
+        content_length = self.headers['Content-Length']
+        if content_length:
+            put_data = self.rfile.read(int(content_length))
+        self.send_response(200)
+        self.send_header('Content-Type', 'text/plain; charset=utf-8')
+        self.send_header('Content-Length', '0')
+        self.send_header('Connection', 'keep-alive')
+        query_components = self.parsed_qs()
+        if 'Key' in query_components and 'Policy' in query_components:
+            self.policies.put(query_components['Key'][0], Policy(int(query_components['Policy'][0])))
+        elif 'uploadId' in query_components and 'partNumber' in query_components:
+            self.send_header('ETag', "SomeTag_" + query_components.get("partNumber")[0])
+        else:
+            self.send_header('ETag', "SomeTag")
+        self.end_headers()
+
+    def do_DELETE(self):
+        query_components = self.parsed_qs()
+        if 'uploadId' in query_components and self.policies.get(
+                query_components.get("uploadId")[0]) == Policy.NONRETRYABLE_FAILURE:
```

Comment:
Please, add comment explaining this if. AFAIU it doesn't try to simulate failure of AbortMultipartUpload, but is rather somehow compensating for previously failed CompleteMultipartUpload?

Reply:
Sorry, my bad, it belong to the branch addressing AbortMultipartUpload [here](https://github.com/kreuzerkrieg/scylladb/tree/EWZ/more-aws-errors-handling). Removing